src/main.cpp

```diff
@@ -260,16 +271,63 @@ struct CNodeState {
     }
 };
 
-/** Map maintaining per-node state. Requires cs_main. */
-map<NodeId, CNodeState> mapNodeState;
+// Class that maintains per-node state, and
+// acts as a RAII smart-pointer that make sure
+// the state stays consistent.
+class NodeStatePtr {
+private:
+    static CCriticalSection cs_mapNodeState;
+    static map<NodeId, CNodeState> mapNodeState;
+    CNodeState* s;
+    NodeId id;
+public:
+    static void insert(NodeId nodeid, const CNode *pnode) {
+        LOCK(cs_mapNodeState);
+        CNodeState &state = mapNodeState.insert(std::make_pair(nodeid, CNodeState())).first->second;
+        state.name = pnode->addrName;
+        state.address = pnode->addr;
+    }
 
-// Requires cs_main.
-CNodeState *State(NodeId pnode) {
-    map<NodeId, CNodeState>::iterator it = mapNodeState.find(pnode);
-    if (it == mapNodeState.end())
-        return NULL;
-    return &it->second;
-}
+    NodeStatePtr(NodeId nodeid) {
+        LOCK(cs_mapNodeState);
+        map<NodeId, CNodeState>::iterator it = mapNodeState.find(nodeid);
+        if (it == mapNodeState.end())
+            s = NULL;
+        else {
+            s = &it->second;
+            id = nodeid;
+            cs_mapNodeState.lock();
+        }
+    }
+    ~NodeStatePtr() {
+        if (s)
+            cs_mapNodeState.unlock();
```

Comment:
It looks like @jtoomim intends for this to be a stack-only object that effectively scope-locks the node state. If someone initializes one of these with `new NodeStatePtr(...)` and saves a reference it may lock the state for longer than intended. But I'm just generally nervous of novel one-off smart pointer implementations.